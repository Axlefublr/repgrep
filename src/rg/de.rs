// NOTE: Originally adapted from the `grep_json_deserialize` crate.
// See: https://github.com/Avi-D-coder/grep_json_deserialize/blob/master/src/lib.rs

use std::ops::Range;

use encoding::{EncoderTrap, EncodingRef};
use serde::{Deserialize, Serialize};

/// A helper to easily select the `RgMessage` kind.
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum RgMessageKind {
  Begin,
  End,
  Match,
  Context,
  Summary,
}

/// A struct used to deserialise JSON values produced by `ripgrep`.
/// See: https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html
#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
#[serde(rename_all = "lowercase")]
#[serde(tag = "type", content = "data")]
pub enum RgMessage {
  /// As specified in: [message-begin](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#message-begin).
  Begin { path: ArbitraryData },
  /// As specified in: [message-end](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#message-end).
  End {
    path: ArbitraryData,
    binary_offset: Option<usize>,
    stats: Stats,
  },
  /// As specified in: [message-match](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#message-match).
  Match {
    path: ArbitraryData,
    lines: ArbitraryData,
    line_number: Option<usize>,
    absolute_offset: usize,
    submatches: Vec<SubMatch>,
  },
  /// As specified in: [message-context](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#message-context).
  Context {
    path: ArbitraryData,
    lines: ArbitraryData,
    line_number: Option<usize>,
    absolute_offset: usize,
    submatches: Vec<SubMatch>,
  },
  Summary {
    elapsed_total: Duration,
    stats: Stats,
  },
}

/// As specified in: [object-arbitrary-data](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#object-arbitrary-data).
#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
#[serde(untagged)]
pub enum ArbitraryData {
  Text { text: String },
  Base64 { bytes: String },
}

impl ArbitraryData {
  pub fn to_vec(&self) -> Vec<u8> {
    match self {
      ArbitraryData::Text { text } => text.as_bytes().to_vec(),
      ArbitraryData::Base64 { bytes } => base64::decode(bytes).unwrap(),
    }
  }

  pub fn to_vec_with_encoding(&self, encoding: EncodingRef) -> Vec<u8> {
    match self {
      ArbitraryData::Text { text } => encoding.encode(text, EncoderTrap::Strict).unwrap(),
      ArbitraryData::Base64 { bytes } => base64::decode(bytes).unwrap(),
    }
  }

  pub fn lossy_utf8(&self) -> String {
    match self {
      ArbitraryData::Text { text } => text.to_owned(),
      ArbitraryData::Base64 { bytes } => {
        String::from_utf8_lossy(base64::decode(bytes).unwrap().as_slice()).to_string()
      }
    }
  }
}

/// As specified in: [object-stats](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#object-stats).
#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
pub struct Stats {
  pub elapsed: Duration,
  pub searches: usize,
  pub searches_with_match: usize,
  pub bytes_searched: usize,
  pub bytes_printed: usize,
  pub matched_lines: usize,
  pub matches: usize,
}

/// As specified in: [object-duration](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#object-duration).
#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
pub struct Duration {
  pub secs: usize,
  pub nanos: usize,
  pub human: String,
}

/// Almost as specified in: [object-submatch](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#object-submatch).
/// `match` is deserialized to `text` because a rust reserves match as a keyword.
#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
#[serde(rename = "submatch")]
pub struct SubMatch {
  #[serde(rename = "match")]
  pub text: ArbitraryData,
  #[serde(flatten)]
  pub range: Range<usize>,
}

#[cfg(test)]
mod tests {
  // tests based on [`grep_printer` example output](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#example)

  use crate::rg::de::{ArbitraryData::*, RgMessage::*, *};

  #[test]
  fn arbitrarydata() {
    let json = r#"{"text":"/home/andrew/sherlock"}"#;
    assert_eq!(
      Text {
        text: "/home/andrew/sherlock".to_owned()
      },
      serde_json::from_str(json).unwrap()
    )
  }

  #[test]
  fn begin_deserialize() {
    let json = r#"{"type":"begin","data":{"path":{"text":"/home/andrew/sherlock"}}}"#;
    assert_eq!(
      Begin {
        path: Text {
          text: "/home/andrew/sherlock".to_owned()
        }
      },
      serde_json::from_str(json).unwrap()
    );
  }

  #[test]
  fn end_deserialize() {
    let json = r#"{"type":"end","data":{"path":{"text":"/home/andrew/sherlock"},"binary_offset":null,"stats":{"elapsed":{"secs":0,"nanos":36296,"human":"0.0000s"},"searches":1,"searches_with_match":1,"bytes_searched":367,"bytes_printed":1151,"matched_lines":2,"matches":2}}}"#;
    assert_eq!(
      End {
        path: Text {
          text: "/home/andrew/sherlock".to_owned()
        },
        binary_offset: None,
        stats: Stats {
          elapsed: Duration {
            secs: 0,
            nanos: 36296,
            human: "0.0000s".to_owned()
          },
          searches: 1,
          searches_with_match: 1,
          bytes_searched: 367,
          bytes_printed: 1151,
          matched_lines: 2,
          matches: 2
        }
      },
      serde_json::from_str(json).unwrap()
    );
  }

  #[test]
  fn match_deserialize() {
    let json = r#"{"type":"match","data":{"path":{"text":"/home/andrew/sherlock"},"lines":{"text":"but Doctor Watson has to have it taken out for him and dusted,\n"},"line_number":5,"absolute_offset":258,"submatches":[{"match":{"text":"Watson"},"start":11,"end":17}]}}"#;
    assert_eq!(
      Match {
        path: Text {
          text: "/home/andrew/sherlock".to_owned()
        },
        lines: Text {
          text: "but Doctor Watson has to have it taken out for him and dusted,\n".to_owned()
        },
        line_number: Some(5),
        absolute_offset: 258,
        submatches: vec![SubMatch {
          text: Text {
            text: "Watson".to_owned()
          },
          range: (11..17)
        }],
      },
      serde_json::from_str(json).unwrap()
    )
  }

  #[test]
  fn content_deserialize() {
    let json = r#"{"type":"context","data":{"path":{"text":"/home/andrew/sherlock"},"lines":{"text":"can extract a clew from a wisp of straw or a flake of cigar ash;\n"},"line_number":4,"absolute_offset":193,"submatches":[]}}"#;
    assert_eq!(
      Context {
        path: Text {
          text: "/home/andrew/sherlock".to_owned()
        },
        lines: Text {
          text: "can extract a clew from a wisp of straw or a flake of cigar ash;\n".to_owned()
        },
        line_number: Some(4),
        absolute_offset: 193,
        submatches: vec![],
      },
      serde_json::from_str(json).unwrap()
    )
  }

  #[test]
  fn summary_deserialize() {
    let json = r#"{"data":{"elapsed_total":{"human":"0.099726s","nanos":99726344,"secs":0},"stats":{"bytes_printed":4106,"bytes_searched":5860,"elapsed":{"human":"0.000047s","nanos":46800,"secs":0},"matched_lines":3,"matches":3,"searches":1,"searches_with_match":1}},"type":"summary"}"#;
    assert_eq!(
      Summary {
        elapsed_total: Duration {
          human: "0.099726s".to_string(),
          nanos: 99_726_344,
          secs: 0
        },
        stats: Stats {
          bytes_printed: 4106,
          bytes_searched: 5860,
          elapsed: Duration {
            human: "0.000047s".to_owned(),
            nanos: 46800,
            secs: 0,
          },
          matched_lines: 3,
          matches: 3,
          searches: 1,
          searches_with_match: 1
        }
      },
      serde_json::from_str(json).unwrap()
    )
  }
}

/// Utilities for tests.
#[cfg(test)]
#[allow(dead_code)]
pub mod test_utilities {
  use crate::rg::de::*;

  impl ArbitraryData {
    pub fn new_with_text(text: String) -> ArbitraryData {
      ArbitraryData::Text { text }
    }

    pub fn new_with_base64(bytes: String) -> ArbitraryData {
      ArbitraryData::Base64 { bytes }
    }
  }

  impl SubMatch {
    pub fn new_text(text: impl AsRef<str>, range: Range<usize>) -> SubMatch {
      SubMatch {
        text: ArbitraryData::new_with_text(text.as_ref().to_owned()),
        range,
      }
    }
    pub fn new_base64(base64: impl AsRef<str>, range: Range<usize>) -> SubMatch {
      SubMatch {
        text: ArbitraryData::new_with_base64(base64.as_ref().to_owned()),
        range,
      }
    }
  }

  impl Duration {
    pub fn new() -> Duration {
      Duration {
        human: String::from("0"),
        nanos: 0,
        secs: 0,
      }
    }
  }

  impl Stats {
    pub fn new() -> Stats {
      Stats {
        bytes_printed: 0,
        bytes_searched: 0,
        matched_lines: 0,
        matches: 0,
        searches: 0,
        searches_with_match: 0,
        elapsed: Duration::new(),
      }
    }
  }

  /// A builder to help construct `RgMessage` structs during tests.
  pub struct RgMessageBuilder {
    kind: RgMessageKind,
    path: Option<ArbitraryData>,
    offset: Option<usize>,
    lines: Option<ArbitraryData>,
    line_number: Option<usize>,
    elapsed_total: Option<Duration>,
    stats: Option<Stats>,
    submatches: Vec<SubMatch>,
  }

  impl RgMessageBuilder {
    pub fn new(kind: RgMessageKind) -> RgMessageBuilder {
      RgMessageBuilder {
        kind,
        path: None,
        offset: None,
        lines: None,
        line_number: None,
        stats: None,
        elapsed_total: None,
        submatches: vec![],
      }
    }

    pub fn with_offset(mut self, offset: usize) -> Self {
      self.offset = Some(offset);
      self
    }

    pub fn with_line_number(mut self, line_number: usize) -> Self {
      self.line_number = Some(line_number);
      self
    }

    pub fn with_path_text(mut self, path: impl AsRef<str>) -> Self {
      self.path = Some(ArbitraryData::new_with_text(path.as_ref().to_owned()));
      self
    }

    pub fn with_lines_text(mut self, lines: impl AsRef<str>) -> Self {
      self.lines = Some(ArbitraryData::new_with_text(lines.as_ref().to_owned()));
      self
    }

    pub fn with_path_base64(mut self, path: impl AsRef<str>) -> Self {
      self.path = Some(ArbitraryData::new_with_base64(path.as_ref().to_owned()));
      self
    }

    pub fn with_lines_base64(mut self, lines: impl AsRef<str>) -> Self {
      self.lines = Some(ArbitraryData::new_with_base64(lines.as_ref().to_owned()));
      self
    }

    pub fn with_submatches(mut self, submatches: Vec<SubMatch>) -> Self {
      self.submatches = submatches;
      self
    }

    pub fn with_elapsed_total(mut self, elapsed_total: Duration) -> Self {
      self.elapsed_total = Some(elapsed_total);
      self
    }

    pub fn with_stats(mut self, stats: Stats) -> Self {
      self.stats = Some(stats);
      self
    }

    pub fn build(self) -> RgMessage {
      match self.kind {
        RgMessageKind::Begin => RgMessage::Begin {
          path: self.path.unwrap(),
        },
        RgMessageKind::End => RgMessage::End {
          path: self.path.unwrap(),
          binary_offset: self.offset,
          stats: self.stats.unwrap(),
        },
        RgMessageKind::Match => RgMessage::Match {
          path: self.path.unwrap(),
          absolute_offset: self.offset.unwrap(),
          line_number: self.line_number,
          lines: self.lines.unwrap(),
          submatches: self.submatches,
        },
        RgMessageKind::Context => RgMessage::Context {
          path: self.path.unwrap(),
          absolute_offset: self.offset.unwrap(),
          line_number: self.line_number,
          lines: self.lines.unwrap(),
          submatches: self.submatches,
        },
        RgMessageKind::Summary => RgMessage::Summary {
          elapsed_total: self.elapsed_total.unwrap(),
          stats: self.stats.unwrap(),
        },
      }
    }
  }
}
