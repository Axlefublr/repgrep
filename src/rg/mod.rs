pub mod de;
pub mod de_printable;
pub mod encoding;
pub mod exec;

pub use self::encoding::*;
