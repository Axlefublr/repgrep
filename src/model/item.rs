use std::ffi::OsString;
use std::path::PathBuf;

use anyhow::Result;
use tui::style::{Color, Style};
use tui::widgets::Text;

use crate::rg::de::{ArbitraryData, RgMessage, RgMessageKind, SubMatch};

// TODO: tests for Base64 decoding on separate platforms

/// Convert Base64 encoded data to an OsString on Unix platforms.
/// https://doc.rust-lang.org/std/ffi/index.html#on-unix
#[cfg(unix)]
fn base64_to_os_string(bytes: Vec<u8>) -> Result<OsString> {
  use std::os::unix::ffi::OsStringExt;
  Ok(OsString::from_vec(bytes))
}

/// Convert Base64 encoded data to an OsString on Windows platforms.
/// https://doc.rust-lang.org/std/ffi/index.html#on-windows
#[cfg(not(unix))]
fn base64_to_os_string(bytes: Vec<u8>) -> Result<OsString> {
  use safe_transmute::{transmute_many, try_copy, PedanticGuard};
  use std::os::windows::ffi::OsStringExt;

  // Transmute decoded Base64 bytes as UTF-16 since that's what underlying paths are on Windows.
  let bytes_u16 = try_copy!(transmute_many::<u16, PedanticGuard>(&bytes))?;
  OsString::from_wide(&bytes_u16)
}

#[derive(Debug, Clone)]
pub struct Item {
  rg_message: RgMessage,

  pub kind: RgMessageKind,
  pub should_replace: bool,
}

impl Item {
  pub fn new(rg_message: RgMessage) -> Item {
    let kind = match rg_message {
      RgMessage::Begin { .. } => RgMessageKind::Begin,
      RgMessage::End { .. } => RgMessageKind::End,
      RgMessage::Match { .. } => RgMessageKind::Match,
      RgMessage::Context { .. } => RgMessageKind::Context,
      RgMessage::Summary { .. } => RgMessageKind::Summary,
    };

    Item {
      rg_message,
      kind,
      should_replace: true,
    }
  }

  pub fn is_selectable(&self) -> bool {
    matches!(self.kind, RgMessageKind::Begin | RgMessageKind::Match)
  }

  pub fn offset(&self) -> Option<usize> {
    match &self.rg_message {
      RgMessage::End { binary_offset, .. } => *binary_offset,
      RgMessage::Match {
        absolute_offset, ..
      } => Some(*absolute_offset),
      _ => None,
    }
  }

  pub fn match_count(&self) -> usize {
    self
      .matches()
      .map(|submatches| submatches.len())
      .unwrap_or(0)
  }

  pub fn matches(&self) -> Option<&[SubMatch]> {
    match &self.rg_message {
      RgMessage::Match { submatches, .. } => Some(submatches),
      _ => None,
    }
  }

  pub fn path(&self) -> Option<PathBuf> {
    let path_data = match &self.rg_message {
      RgMessage::Begin { path, .. } => path,
      RgMessage::Match { path, .. } => path,
      RgMessage::Context { path, .. } => path,
      RgMessage::End { path, .. } => path,
      RgMessage::Summary { .. } => return None,
    };

    Some(match path_data {
      ArbitraryData::Text { text } => PathBuf::from(text),
      ArbitraryData::Base64 { bytes } => {
        // Decode the Base64 into u8 bytes.
        let data = match base64::decode(bytes) {
          Ok(data) => data,
          Err(e) => panic!("Error deserialising Base64 data: {}", e),
        };

        // Convert the bytes into an OsString.
        let os_string = match base64_to_os_string(data) {
          Ok(os_string) => os_string,
          Err(e) => panic!("Error transmuting Base64 data to OsString: {}", e),
        };

        PathBuf::from(os_string)
      }
    })
  }

  pub fn to_text(&self, replacement: Option<&str>) -> Text {
    // TODO: handle multiline matches
    match &self.rg_message {
      RgMessage::Begin { .. } => Text::styled(
        format!("{}", self.path().unwrap().display()),
        Style::default().fg(Color::Magenta),
      ),
      RgMessage::Context {
        lines, line_number, ..
      } => {
        let mut text = lines.lossy_utf8();
        if let Some(number) = line_number {
          text = format!("{}:{}", number, text);
        }

        Text::styled(text, Style::default().fg(Color::DarkGray))
      }
      RgMessage::Match {
        lines,
        line_number,
        submatches,
        ..
      } => {
        // TODO: highlight matches (red) on line and replacements (green). Currently not possible.
        // See: https://github.com/fdehau/tui-rs/issues/315
        let mut style = Style::default();
        if !self.should_replace {
          style = style.fg(Color::Red);
        }

        // TODO: when we can highlight mid-text, don't replace the match, colour the match (submatch.text.lossy_utf8())
        // and add the replacement after.
        // If we have a replacement, then perform the replacement on the bytes before encoding as UTF8.
        let mut bytes = lines.to_vec();
        if self.should_replace {
          if let Some(replacement) = replacement {
            let replacement = replacement.as_bytes().to_vec();
            for submatch in submatches.iter().rev() {
              bytes.splice(submatch.range.clone(), replacement.clone());
            }
          }
        }

        // Prepend the line number.
        let mut text: String = String::from_utf8_lossy(&bytes).to_string();
        if let Some(number) = line_number {
          text = format!("{}:{}", number, text);
        }

        Text::styled(text, style)
      }
      RgMessage::End { .. } => Text::raw(""),
      RgMessage::Summary { elapsed_total, .. } => {
        Text::raw(format!("Search duration: {}", elapsed_total.human))
      }
    }
  }
}

#[cfg(test)]
mod tests {
  use std::path::PathBuf;

  use pretty_assertions::assert_eq;
  use tui::style::{Color, Style};
  use tui::widgets::Text;

  use crate::model::*;
  use crate::rg::de::test_utilities::*;
  use crate::rg::de::*;

  const RG_JSON_BEGIN: &str = r#"{"type":"begin","data":{"path":{"text":"src/model/item.rs"}}}"#;
  const RG_JSON_MATCH: &str = r#"{"type":"match","data":{"path":{"text":"src/model/item.rs"},"lines":{"text":"    Item::new(rg_msg)\n"},"line_number":197,"absolute_offset":5522,"submatches":[{"match":{"text":"rg_msg"},"start":14,"end":20}]}}"#;
  const RG_JSON_CONTEXT: &str = r#"{"type":"context","data":{"path":{"text":"src/model/item.rs"},"lines":{"text":"  }\n"},"line_number":198,"absolute_offset":5544,"submatches":[]}}"#;
  const RG_JSON_END: &str = r#"{"type":"end","data":{"path":{"text":"src/model/item.rs"},"binary_offset":null,"stats":{"elapsed":{"secs":0,"nanos":97924,"human":"0.000098s"},"searches":1,"searches_with_match":1,"bytes_searched":5956,"bytes_printed":674,"matched_lines":2,"matches":2}}}"#;
  const RG_JSON_SUMMARY: &str = r#"{"data":{"elapsed_total":{"human":"0.013911s","nanos":13911027,"secs":0},"stats":{"bytes_printed":3248,"bytes_searched":18789,"elapsed":{"human":"0.000260s","nanos":260276,"secs":0},"matched_lines":10,"matches":10,"searches":2,"searches_with_match":2}},"type":"summary"}"#;

  fn new_item(raw_json: &str) -> Item {
    let rg_msg = serde_json::from_str::<RgMessage>(raw_json).unwrap();
    Item::new(rg_msg)
  }

  #[test]
  fn item_kind_matches_rg_message_kind() {
    assert_eq!(new_item(RG_JSON_BEGIN).kind, RgMessageKind::Begin);
    assert_eq!(new_item(RG_JSON_MATCH).kind, RgMessageKind::Match);
    assert_eq!(new_item(RG_JSON_CONTEXT).kind, RgMessageKind::Context);
    assert_eq!(new_item(RG_JSON_END).kind, RgMessageKind::End);
    assert_eq!(new_item(RG_JSON_SUMMARY).kind, RgMessageKind::Summary);
  }

  #[test]
  fn only_match_and_begin_are_selectable() {
    assert_eq!(new_item(RG_JSON_BEGIN).is_selectable(), true);
    assert_eq!(new_item(RG_JSON_MATCH).is_selectable(), true);
    assert_eq!(new_item(RG_JSON_CONTEXT).is_selectable(), false);
    assert_eq!(new_item(RG_JSON_END).is_selectable(), false);
    assert_eq!(new_item(RG_JSON_SUMMARY).is_selectable(), false);
  }

  #[test]
  fn match_count() {
    assert_eq!(new_item(RG_JSON_BEGIN).match_count(), 0);
    assert_eq!(new_item(RG_JSON_MATCH).match_count(), 1);
    assert_eq!(new_item(RG_JSON_CONTEXT).match_count(), 0);
    assert_eq!(new_item(RG_JSON_END).match_count(), 0);
    assert_eq!(new_item(RG_JSON_SUMMARY).match_count(), 0);
  }

  #[test]
  fn matches() {
    assert_eq!(new_item(RG_JSON_BEGIN).matches(), None);
    assert_eq!(
      new_item(RG_JSON_MATCH).matches(),
      Some([SubMatch::new_text("rg_msg", 14..20)].as_ref())
    );
    assert_eq!(new_item(RG_JSON_CONTEXT).matches(), None);
    assert_eq!(new_item(RG_JSON_END).matches(), None);
    assert_eq!(new_item(RG_JSON_SUMMARY).matches(), None);
  }

  #[test]
  fn offset() {
    assert_eq!(new_item(RG_JSON_BEGIN).offset(), None);
    assert_eq!(new_item(RG_JSON_MATCH).offset(), Some(5522));
    assert_eq!(new_item(RG_JSON_CONTEXT).offset(), None);
    assert_eq!(new_item(RG_JSON_END).offset(), None);
    assert_eq!(new_item(RG_JSON_SUMMARY).offset(), None);
  }

  #[test]
  fn binary_offset() {
    let item = new_item(
      r#"{"type":"end","data":{"path":{"text":"src/model/item.rs"},"binary_offset":1234,"stats":{"elapsed":{"secs":0,"nanos":97924,"human":"0.000098s"},"searches":1,"searches_with_match":1,"bytes_searched":5956,"bytes_printed":674,"matched_lines":2,"matches":2}}}"#,
    );
    assert_eq!(item.offset(), Some(1234));
  }

  #[test]
  fn path_with_text() {
    let path = PathBuf::from("src/model/item.rs");
    assert_eq!(new_item(RG_JSON_BEGIN).path().as_ref(), Some(&path));
    assert_eq!(new_item(RG_JSON_MATCH).path().as_ref(), Some(&path));
    assert_eq!(new_item(RG_JSON_CONTEXT).path().as_ref(), Some(&path));
    assert_eq!(new_item(RG_JSON_END).path().as_ref(), Some(&path));
    assert_eq!(new_item(RG_JSON_SUMMARY).path().as_ref(), None);
  }

  // TODO: write a similar test for Windows systems
  #[test]
  #[cfg(unix)]
  fn path_with_base64() {
    use std::ffi::OsStr;
    use std::os::unix::ffi::OsStrExt;

    // Here, the values 0x66 and 0x6f correspond to 'f' and 'o'
    // respectively. The value 0x80 is a lone continuation byte, invalid
    // in a UTF-8 sequence.
    let invalid_utf8_name_bytes = [0x66, 0x6f, 0x80, 0x6f];
    let invalid_utf8_name = OsStr::from_bytes(&invalid_utf8_name_bytes[..]);
    let invalid_utf8_path = PathBuf::from(invalid_utf8_name);

    let new_item_path_base64 = |kind| {
      Item::new(
        RgMessageBuilder::new(kind)
          .with_path_base64(base64::encode(&invalid_utf8_name_bytes))
          .with_lines_text("foo bar baz")
          .with_submatches(vec![SubMatch::new_text("foo", 0..3)])
          .with_stats(Stats::new())
          .with_elapsed_total(Duration::new())
          .with_offset(0)
          .build(),
      )
    };

    assert_eq!(
      new_item_path_base64(RgMessageKind::Begin).path().as_ref(),
      Some(&invalid_utf8_path)
    );
    assert_eq!(
      new_item_path_base64(RgMessageKind::Match).path().as_ref(),
      Some(&invalid_utf8_path)
    );
    assert_eq!(
      new_item_path_base64(RgMessageKind::Context).path().as_ref(),
      Some(&invalid_utf8_path)
    );
    assert_eq!(
      new_item_path_base64(RgMessageKind::End).path().as_ref(),
      Some(&invalid_utf8_path)
    );
    assert_eq!(
      new_item_path_base64(RgMessageKind::Summary).path().as_ref(),
      None
    );
  }

  #[test]
  fn to_text_with_text() {
    let s = Style::default();

    // Without replacement.
    assert_eq!(
      new_item(RG_JSON_BEGIN).to_text(None),
      Text::styled("src/model/item.rs", s.fg(Color::Magenta))
    );
    assert_eq!(
      new_item(RG_JSON_MATCH).to_text(None),
      Text::styled("197:    Item::new(rg_msg)\n", s)
    );
    assert_eq!(
      new_item(RG_JSON_CONTEXT).to_text(None),
      Text::styled("198:  }\n", s.fg(Color::DarkGray))
    );
    assert_eq!(new_item(RG_JSON_END).to_text(None), Text::raw(""));
    assert_eq!(
      new_item(RG_JSON_SUMMARY).to_text(None),
      Text::raw("Search duration: 0.013911s")
    );

    // With replacement.
    let replacement = "foobar";
    assert_eq!(
      new_item(RG_JSON_BEGIN).to_text(Some(replacement)),
      Text::styled("src/model/item.rs", s.fg(Color::Magenta))
    );
    assert_eq!(
      new_item(RG_JSON_MATCH).to_text(Some(replacement)),
      Text::styled("197:    Item::new(foobar)\n", s)
    );
    assert_eq!(
      new_item(RG_JSON_CONTEXT).to_text(Some(replacement)),
      Text::styled("198:  }\n", s.fg(Color::DarkGray))
    );
    assert_eq!(
      new_item(RG_JSON_END).to_text(Some(replacement)),
      Text::raw("")
    );
    assert_eq!(
      new_item(RG_JSON_SUMMARY).to_text(Some(replacement)),
      Text::raw("Search duration: 0.013911s")
    );
  }

  #[test]
  fn to_text_with_base64_lossy() {
    // The following types are skipped because:
    // Begin:   already tested via the `path_with_base64` test.
    // End:     already tested via the `path_with_base64` test.
    // Summary: doesn't include an `ArbitraryData` struct.

    let b64_json_match = r#"{"type":"match","data":{"path":{"text":"src/model/item.rs"},"lines":{"bytes":"ICAgIEl0ZW06Ov9uZXcocmdfbXNnKQo="},"line_number":197,"absolute_offset":5522,"submatches":[{"match":{"text":"rg_msg"},"start":15,"end":21}]}}"#;
    let b64_json_context = r#"{"type":"context","data":{"path":{"text":"src/model/item.rs"},"lines":{"bytes":"ICD/fQo="},"line_number":198,"absolute_offset":5544,"submatches":[]}}"#;

    // Since we don't read the entire file when we view the results, we expect the UTF8 replacement character.
    // Without replacement.
    let s = Style::default();
    assert_eq!(
      new_item(b64_json_match).to_text(None),
      Text::styled("197:    Item::�new(rg_msg)\n", s)
    );
    assert_eq!(
      new_item(b64_json_context).to_text(None),
      Text::styled("198:  �}\n", s.fg(Color::DarkGray))
    );

    // With replacement.
    let replacement = "foobar";
    assert_eq!(
      new_item(b64_json_match).to_text(Some(replacement)),
      Text::styled("197:    Item::�new(foobar)\n", s)
    );
    assert_eq!(
      new_item(b64_json_context).to_text(Some(replacement)),
      Text::styled("198:  �}\n", s.fg(Color::DarkGray))
    );
  }
}
