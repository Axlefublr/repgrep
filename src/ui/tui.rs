use std::io::{self, Write};
use std::sync::mpsc;
use std::thread;

use anyhow::{anyhow, Result};
use crossterm::event::{self, DisableMouseCapture, EnableMouseCapture};
use crossterm::execute;
use crossterm::terminal::{self, EnterAlternateScreen, LeaveAlternateScreen};
use tui::{backend::CrosstermBackend, Terminal};

use crate::model::ReplacementCriteria;
use crate::rg::de::RgMessage;
use crate::ui::app::{App, AppState};

const MINIMUM_WIDTH: u16 = 40;
const MINIMUM_HEIGHT: u16 = 40;

pub struct Tui {
    app: App,
}

impl Tui {
    pub fn new(rg_cmdline: String, rg_results: Vec<RgMessage>) -> Tui {
        Tui {
            app: App::new(rg_cmdline, rg_results),
        }
    }

    pub fn start(mut self) -> Result<Option<ReplacementCriteria>> {
        terminal::enable_raw_mode()?;

        let mut stdout = io::stdout();
        execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;

        let backend = CrosstermBackend::new(stdout);
        let mut term = Terminal::new(backend)?;
        term.hide_cursor()?;

        // Setup input handling
        let (tx, rx) = mpsc::channel();

        thread::spawn(move || loop {
            match tx.send(event::read().expect("failed to read event from terminal")) {
                Ok(_) => {}
                Err(e) => log::warn!("failed to send event to the main thread: {}", e),
            }
        });

        term.clear()?;

        loop {
            let term_size = term.size()?;
            if term_size.width < MINIMUM_WIDTH || term_size.height < MINIMUM_HEIGHT {
                return Err(anyhow!(
                    "Minimum terminal dimensions are {}x{}!",
                    MINIMUM_WIDTH,
                    MINIMUM_HEIGHT
                ));
            }

            term.draw(|mut f| self.app.draw(&mut f))?;

            let event = rx.recv()?;
            self.app.on_event(term_size, event)?;

            match self.app.state {
                AppState::Running => continue,
                AppState::Cancelled => return Ok(None),
                AppState::Complete(replacement_criteria) => return Ok(Some(replacement_criteria)),
            }
        }
    }

    pub fn restore_terminal() -> Result<()> {
        let backend = CrosstermBackend::new(io::stdout());
        let mut term = Terminal::new(backend)?;

        terminal::disable_raw_mode()?;
        execute!(
            term.backend_mut(),
            LeaveAlternateScreen,
            DisableMouseCapture
        )?;
        term.show_cursor()?;
        term.clear()?;
        term.set_cursor(0, 0)?;

        Ok(())
    }
}
