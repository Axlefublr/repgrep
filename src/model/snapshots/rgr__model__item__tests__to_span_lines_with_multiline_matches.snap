---
source: src/model/item.rs
expression: new_item(RG_JSON_MATCH_MULTILINE).to_span_lines(&ctx)
---
[
    Spans(
        [
            Span {
                content: "3:",
                style: Style {
                    fg: Some(
                        DarkGray,
                    ),
                    bg: None,
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
            Span {
                content: "baz ",
                style: Style {
                    fg: None,
                    bg: None,
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
            Span {
                content: "1",
                style: Style {
                    fg: Some(
                        Black,
                    ),
                    bg: Some(
                        Red,
                    ),
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
        ],
    ),
    Spans(
        [
            Span {
                content: "4:",
                style: Style {
                    fg: Some(
                        DarkGray,
                    ),
                    bg: None,
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
            Span {
                content: "22",
                style: Style {
                    fg: Some(
                        Black,
                    ),
                    bg: Some(
                        Red,
                    ),
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
        ],
    ),
    Spans(
        [
            Span {
                content: "5:",
                style: Style {
                    fg: Some(
                        DarkGray,
                    ),
                    bg: None,
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
            Span {
                content: "333",
                style: Style {
                    fg: Some(
                        Black,
                    ),
                    bg: Some(
                        Red,
                    ),
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
            Span {
                content: " bar ",
                style: Style {
                    fg: None,
                    bg: None,
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
            Span {
                content: "4444",
                style: Style {
                    fg: Some(
                        Black,
                    ),
                    bg: Some(
                        Red,
                    ),
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
            Span {
                content: "",
                style: Style {
                    fg: None,
                    bg: None,
                    add_modifier: (empty),
                    sub_modifier: (empty),
                },
            },
        ],
    ),
]
